federation/pkg/federation-controller/service/dns.go

```diff
@@ -16,25 +16,230 @@ limitations under the License.
 
 package service
 
+import (
+	"fmt"
+	"k8s.io/kubernetes/federation/pkg/dnsprovider"
+	"k8s.io/kubernetes/federation/pkg/dnsprovider/rrstype"
+)
+
+const (
+	// minDnsTtl is the minimum safe DNS TTL value to use (in seconds).  Less than this, and some caching DNS servers might ignore it.
+	minDnsTtl = 180
+)
+
+// getHealthyEndpoints returns the hostnames and/or IP addresses of healthy endpoints for the service, at a zone, region and global level (or an error)
+func (s *ServiceController) getHealthyEndpoints(clusterName string, cachedService *cachedService) (zoneEndpoints, regionEndpoints, globalEndpoints []string, err error) {
+	zoneEndpoints = []string{}
+	regionEndpoints = []string{}
+	globalEndpoints = []string{}
+	zoneName, err := s.getClusterZoneName(clusterName)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	regionName, err := s.getClusterRegionName(clusterName)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	for lbClusterName, lbStatus := range cachedService.serviceStatusMap {
+		lbZoneName, err := s.getClusterZoneName(lbClusterName)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		lbRegionName, err := s.getClusterRegionName(lbClusterName)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		for _, ingress := range lbStatus.Ingress {
+			var address string
+			// We should get either an IP address or a hostname - use whichever one we get
+			if ingress.IP != "" {
+				address = ingress.IP
+			}
+			if ingress.Hostname != "" {
```

Comment:
This should be `else if` of the previous `if` otherwise it overwrites the IP address.